diff --git a/lua-nginx-module/src/api/ngx_http_lua_api.h b/lua-nginx-module/src/api/ngx_http_lua_api.h
index 0181849..c3fa37d 100644
--- a/lua-nginx-module/src/api/ngx_http_lua_api.h
+++ b/lua-nginx-module/src/api/ngx_http_lua_api.h
@@ -22,7 +22,10 @@
 #define ngx_http_lua_version  10024
 
 
+#ifndef NGX_HTTP_LUA_CO_CTX_T_DEFINED
+#define NGX_HTTP_LUA_CO_CTX_T_DEFINED 1
 typedef struct ngx_http_lua_co_ctx_s  ngx_http_lua_co_ctx_t;
+#endif
 
 
 typedef struct {
diff --git a/lua-nginx-module/src/ngx_http_lua_common.h b/lua-nginx-module/src/ngx_http_lua_common.h
index d977ee8..3d2cd6f 100644
--- a/lua-nginx-module/src/ngx_http_lua_common.h
+++ b/lua-nginx-module/src/ngx_http_lua_common.h
@@ -157,7 +157,10 @@ typedef struct {
 #endif
 
 
+#ifndef NGX_HTTP_LUA_CO_CTX_T_DEFINED
+#define NGX_HTTP_LUA_CO_CTX_T_DEFINED 1
 typedef struct ngx_http_lua_co_ctx_s  ngx_http_lua_co_ctx_t;
+#endif
 
 typedef struct ngx_http_lua_sema_mm_s  ngx_http_lua_sema_mm_t;
 
