diff -ruN nginx-rtmp-module-master.orig/config nginx-rtmp-module-master/config
--- nginx-rtmp-module-master.orig/config	2015-03-23 19:30:37.000000000 +0000
+++ nginx-rtmp-module-master/config	2016-02-13 03:00:39.730212206 +0000
@@ -1,6 +1,5 @@
 ngx_addon_name="ngx_rtmp_module"
-
-CORE_MODULES="$CORE_MODULES
+RTMP_CORE_MODULES="                                         \
                 ngx_rtmp_module                             \
                 ngx_rtmp_core_module                        \
                 ngx_rtmp_cmd_module                         \
@@ -21,15 +20,11 @@
                 ngx_rtmp_hls_module                         \
                 ngx_rtmp_dash_module                        \
                 "
-
-
-HTTP_MODULES="$HTTP_MODULES                                 \
+RTMP_HTTP_MODULES="                                         \
                 ngx_rtmp_stat_module                        \
                 ngx_rtmp_control_module                     \
                 "
-
-
-NGX_ADDON_DEPS="$NGX_ADDON_DEPS                             \
+RTMP_DEPS="                                                 \
                 $ngx_addon_dir/ngx_rtmp_amf.h               \
                 $ngx_addon_dir/ngx_rtmp_bandwidth.h         \
                 $ngx_addon_dir/ngx_rtmp_cmd_module.h        \
@@ -48,9 +43,7 @@
                 $ngx_addon_dir/hls/ngx_rtmp_mpegts.h        \
                 $ngx_addon_dir/dash/ngx_rtmp_mp4.h          \
                 "
-
-
-NGX_ADDON_SRCS="$NGX_ADDON_SRCS                             \
+RTMP_CORE_SRCS="                                            \
                 $ngx_addon_dir/ngx_rtmp.c                   \
                 $ngx_addon_dir/ngx_rtmp_init.c              \
                 $ngx_addon_dir/ngx_rtmp_handshake.c         \
@@ -70,8 +63,6 @@
                 $ngx_addon_dir/ngx_rtmp_flv_module.c        \
                 $ngx_addon_dir/ngx_rtmp_mp4_module.c        \
                 $ngx_addon_dir/ngx_rtmp_netcall_module.c    \
-                $ngx_addon_dir/ngx_rtmp_stat_module.c       \
-                $ngx_addon_dir/ngx_rtmp_control_module.c    \
                 $ngx_addon_dir/ngx_rtmp_relay_module.c      \
                 $ngx_addon_dir/ngx_rtmp_bandwidth.c         \
                 $ngx_addon_dir/ngx_rtmp_exec_module.c       \
@@ -86,7 +77,26 @@
                 $ngx_addon_dir/hls/ngx_rtmp_mpegts.c        \
                 $ngx_addon_dir/dash/ngx_rtmp_mp4.c          \
                 "
-CFLAGS="$CFLAGS -I$ngx_addon_dir"
+RTMP_HTTP_SRCS="                                            \
+                $ngx_addon_dir/ngx_rtmp_stat_module.c       \
+                $ngx_addon_dir/ngx_rtmp_control_module.c    \
+                "
+ngx_module_incs=$ngx_addon_dir
+ngx_module_deps=$RTMP_DEPS
 
-USE_OPENSSL=YES
+if [ $ngx_module_link = DYNAMIC ] ; then
+    ngx_module_name="$RTMP_CORE_MODULES $RTMP_HTTP_MODULES"
+    ngx_module_srcs="$RTMP_CORE_SRCS $RTMP_HTTP_SRCS"
+    . auto/module
+elif [ $ngx_module_link = ADDON ] ; then
+    ngx_module_type=CORE
+    ngx_module_name=$RTMP_CORE_MODULES
+    ngx_module_srcs=$RTMP_CORE_SRCS
+    . auto/module
+    ngx_module_type=HTTP
+    ngx_module_name=$RTMP_HTTP_MODULES
+    ngx_module_srcs=$RTMP_HTTP_SRCS
+    . auto/module
+fi
 
+USE_OPENSSL=YES
