diff -ruN ngx_http_secure_download-master.orig/config ngx_http_secure_download-master/config
--- ngx_http_secure_download-master.orig/config	2014-04-17 07:24:54.000000000 +0000
+++ ngx_http_secure_download-master/config	2016-02-11 23:00:40.326772371 +0000
@@ -1,5 +1,16 @@
 USE_SHA1=YES
 ngx_addon_name=ngx_http_secure_download
-HTTP_MODULES="$HTTP_MODULES ngx_http_secure_download_module"
-NGX_ADDON_SRCS="$NGX_ADDON_SRCS $ngx_addon_dir/ngx_http_secure_download_module.c"
-CORE_LIBS="$CORE_LIBS -lmhash"
+HTTP_SECURE_DOWNLOAD_SRCS="$ngx_addon_dir/ngx_http_secure_download_module.c"
+HTTP_SECURE_DOWNLOAD_LIBS="-lmhash"
+if test -n "$ngx_module_link"; then
+    ngx_module_type=HTTP
+    ngx_module_name="$ngx_addon_name"
+    ngx_module_srcs="$HTTP_SECURE_DOWNLOAD_SRCS"
+    ngx_module_libs="$HTTP_SECURE_DOWNLOAD_LIBS"
+
+    . auto/module
+else
+    HTTP_MODULES="$HTTP_MODULES $ngx_addon_name"
+    NGX_ADDON_SRCS="$NGX_ADDON_SRCS $HTTP_SECURE_DOWNLOAD_SRCS"
+    CORE_LIBS="$CORE_LIBS $HTTP_SECURE_DOWNLOAD_LIBS"
+fi
